rust/src/bs_safe.rs

```diff
@@ -0,0 +1,610 @@
+use crate::ffi;
+use std::borrow::Borrow;
+use std::collections::hash_map::DefaultHasher;
+use std::ffi::c_char;
+use std::ffi::c_int;
+use std::ffi::c_uint;
+use std::ffi::CStr;
+use std::ffi::OsStr;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::mem::forget;
+use std::mem::ManuallyDrop;
+use std::ptr::NonNull;
+use std::ptr::{null, null_mut};
+use std::str::Utf8Error;
+use std::time::Instant;
+
+// This is more or less equivalent to manually defining Display and From<other error types>
+use thiserror::Error;
+
+/// A loaded shared library for a stan model
+pub struct StanLibrary {
+    lib: ManuallyDrop<ffi::BridgeStan>,
+    id: u64,
+}
+
+// To work around a bug on windows where unloading a library
+// can lead to deadlocks.
+//
+// See https://github.com/roualdes/bridgestan/issues/111
+impl Drop for StanLibrary {
+    fn drop(&mut self) {
+        let lib = unsafe { ManuallyDrop::take(&mut self.lib) };
+        forget(lib.into_library());
+    }
+}
+
+/// A callback for print statements in stan models
+pub type StanPrintCallback = extern "C" fn(*const c_char, usize);
+
+impl StanLibrary {
+    /// Provide a callback function to be called when stan prints a message
+    ///
+    /// # Safety
+    ///
+    /// The provided function must never panic.
+    ///
+    /// Since the call is proteted by a mutex internally, it does not
+    /// need to be thread safe.
+    pub unsafe fn set_print_callback(&mut self, callback: StanPrintCallback) -> Result<()> {
+        let mut err = ErrorMsg::new(self);
+        let rc = unsafe { self.lib.bs_set_print_callback(Some(callback), err.as_ptr()) };
+
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("Could not load target library: {0}")]
+pub struct LoadingError(#[from] libloading::Error);
+
+/// Error type for bridgestan interface
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum BridgeStanError {
+    #[error(transparent)]
+    InvalidLibrary(#[from] LoadingError),
+    #[error("Bad Stan library version: Got {0} but expected {1}")]
+    BadLibraryVersion(String, String),
+    #[error("The Stan library was compiled without threading support. Config was {0}")]
+    StanThreads(String),
+    #[error("Failed to decode string to UTF8")]
+    InvalidString(#[from] Utf8Error),
+    #[error("Failed to construct model: {0}")]
+    ConstructFailed(String),
+    #[error("Failed during evaluation: {0}")]
+    EvaluationFailed(String),
```

Comment:
I noticed we're using this enum variant for things which aren't strictly "evaluation" (e.g. a failure in `set_print_callback`). Should we split it up or rename it?

Reply:
I made a separate one for `set_print_callback`. I guess we could also make another one for the (un)constrain param functions, I'm not sure if that would be better or not.